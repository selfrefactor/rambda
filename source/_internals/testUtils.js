import combinate from 'combinate'
import {
  equals,
  filter,
  forEach,
  init,
  last,
  omit,
  switcher,
  type,
} from 'rambdax'

const omitOk = omit('ok')

const PENDING = 'PENDING'
const RESULTS_EQUAL = 'results are equal'
const RESULTS_MISMATCH = 'results are different'
const ERRORS_EQUAL = 'errors are equal'
const ERRORS_MESSAGE_MISMATCH = 'errors messages are different'
const ERRORS_TYPE_MISMATCH = 'errors types are different'
const SHOULD_THROW = 'Rambda should throw'
const SHOULD_NOT_THROW = 'Rambda should not throw'
const ALL_ERROR_LABELS = {
  RESULTS_MISMATCH,
  ERRORS_MESSAGE_MISMATCH,
  ERRORS_TYPE_MISMATCH,
  SHOULD_NOT_THROW,
  SHOULD_THROW,
}
const MISSING = 'Missing error handle in parseError'

function parseError(err){
  const typeError = switcher(err)
    .is(x => x instanceof TypeError, 'TypeError')
    .is(x => x instanceof SyntaxError, 'SyntaxError')
    .is(x => x instanceof RangeError, 'RangeError')
    .is(x => x instanceof Error, 'Error')
    .default(MISSING)

  if (typeError === MISSING){
    throw new Error('typeError === MISSING')
  }

  return {
    message : err.message,
    type    : typeError,
    ok      : true,
  }
}

function executeSync(
  fn, inputs, returnsFunctionFlag
){
  let result = PENDING
  let error = { ok : false }
  try {
    result = returnsFunctionFlag ?
      fn(...init(inputs))(last(inputs)) :
      fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

async function executeAsync(
  fn, inputs, returnsFunctionFlag
){
  let result = PENDING
  let error = { ok : false }
  try {
    result = returnsFunctionFlag ?
      await fn(...init(inputs))(last(inputs)) :
      await fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

export function profileMethod({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  fn,
}){
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const { result, error } = executeSync(
        fn, inputs, returnsFunctionFlag
      )

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function profileMethodAsync({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  fn,
}){
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)
    test(getTestTitle(...inputs), async () => {
      const { result, error } = await executeAsync(
        fn,
        inputs,
        returnsFunctionFlag
      )

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function compareToRamda(
  fn, fnRamda, returnsFunctionFlag
){
  return (...inputs) => {
    const { result, error } = executeSync(
      fn, inputs, returnsFunctionFlag
    )
    const { result: ramdaResult, error: ramdaError } = executeSync(
      fnRamda,
      inputs,
      returnsFunctionFlag
    )

    const toReturn = {
      result,
      ramdaResult,
      ramdaError : ramdaError.ok ? omitOk(ramdaError) : PENDING,
      error      : error.ok ? omitOk(error) : PENDING,
    }

    if (result !== PENDING){
      if (ramdaError.ok){
        return {
          ...toReturn,
          ok    : false,
          label : SHOULD_THROW,
        }
      }

      if (equals(result, ramdaResult)){
        return {
          ...toReturn,
          ok    : true,
          label : RESULTS_EQUAL,
        }
      }

      return {
        ...toReturn,
        ok    : false,
        label : RESULTS_MISMATCH,
      }
    }

    if (equals(error, ramdaError)){
      return {
        ...toReturn,
        ok    : true,
        label : ERRORS_EQUAL,
      }
    }

    if (ramdaError.ok){
      return {
        ...toReturn,
        ok    : false,
        label :
          ramdaError.type === error.type ?
            ERRORS_MESSAGE_MISMATCH :
            ERRORS_TYPE_MISMATCH,
      }
    }

    return {
      ...toReturn,
      ok    : false,
      label : SHOULD_NOT_THROW,
    }
  }
}

export const getTestTitle = (...inputs) => inputs.map(type).join(' | ')

export const compareCombinations = ({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  setCounter = () => {},
  callback = x => {},
  fn,
  fnRamda,
}) => {
  const counter = {
    RESULTS_MISMATCH        : 0,
    SHOULD_THROW            : 0,
    SHOULD_NOT_THROW        : 0,
    ERRORS_TYPE_MISMATCH    : 0,
    ERRORS_MESSAGE_MISMATCH : 0,
    TOTAL_TESTS             : 0,
  }

  const increaseCounter = comparedResult => {
    if (comparedResult.ok) return
    let counterProp
    forEach((x, prop) => {
      if (x === comparedResult.label) counterProp = prop
    }, ALL_ERROR_LABELS)
    counter[ counterProp ]++
  }

  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)
  const compareOutputs = compareToRamda(
    fn, fnRamda, returnsFunctionFlag
  )

  afterAll(() => callback(counter))

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const compared = compareOutputs(...inputs)
      setCounter()
      if (!compared.ok){
        // if (compared.label === RESULTS_MISMATCH){
        increaseCounter(compared)
        expect({
          ...compared,
          inputs,
        }).toMatchSnapshot()
      }
      counter.TOTAL_TESTS++
    })
  })
}

const demoMap = new Map()

demoMap.set('a', 1)
demoMap.set('b', 2)

const demoWeakMap = new WeakMap()

demoWeakMap.set({ a : 0 }, 1)
demoWeakMap.set({ b : 0 }, 2)

const demoWeakSet = new WeakSet()
const foo = { a : 1 }
const bar = { b : 2 }

demoWeakSet.add(foo)
demoWeakSet.add(bar)

const demoBigInt = BigInt(9007199254740991)

const demoRegExp = new RegExp('ab+c', 'i')
const demoSymbol = Symbol('foo')

export const FALSY_VALUES = [ null, undefined, false, NaN, '' ]
export const EXTRA_BUILD_IN_OBJECTS = [
  Infinity,
  -Infinity,
  Symbol,
  new ArrayBuffer(8),
  demoRegExp,
  demoMap,
  demoWeakMap,
  demoWeakSet,
  demoBigInt,
  demoSymbol,
]
