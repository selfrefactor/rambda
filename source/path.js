export function path(list, obj){
  if (arguments.length === 1) return _obj => path(list, _obj)

  if (obj === null || obj === undefined){
    return undefined
  }
  let willReturn = obj
  let counter = 0

  const pathArrValue = typeof list === 'string' ? list.split('.') : list

  while (counter < pathArrValue.length){
    if (willReturn === null || willReturn === undefined){
      return undefined
    }
    willReturn = willReturn[ pathArrValue[ counter ] ]
    counter++
  }

  return willReturn
}
