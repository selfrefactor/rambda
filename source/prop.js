export function prop(key, obj){
  if (arguments.length === 1) return _obj => prop(key, _obj)

  if (!obj) return undefined

  return obj[ key ]
}
