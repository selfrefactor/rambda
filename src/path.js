import { createPath } from './_internals/createPath.js'

export function pathFn(pathInput, obj){
  let willReturn = obj
  let counter = 0

  const pathArrValue = createPath(pathInput)

  while (counter < pathArrValue.length){
    if (willReturn === null || willReturn === undefined){
      return undefined
    }
    if (willReturn[ pathArrValue[ counter ] ] === null) return undefined

    willReturn = willReturn[ pathArrValue[ counter ] ]
    counter++
  }

  return willReturn
}

export function path(pathInput, obj){
  if (arguments.length === 1) return _obj => path(pathInput, _obj)

  if (obj === null || obj === undefined){
    return undefined
  }

  return pathFn(pathInput, obj)
}
