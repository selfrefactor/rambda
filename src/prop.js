/**
 * Returns a function that when supplied an object returns the indicated
 * property of that object, if it exists.
 *
 * @func
 * @category Object
 * @sig s -> {s: a} -> a | Undefined
 * @param {String} key The property name
 * @param {Object} obj The object to query
 * @return {*} The value at `obj[key]`.
 * @example
 *
 *      R.prop('x', {x: 100}); //=> 100
 *      R.prop('x', {}); //=> undefined
 *      R.compose(R.inc, R.prop('x'))({ x: 3 }) //=> 4
 */
export function prop(key, obj){
  if (arguments.length === 1) return _obj => prop(key, _obj)

  if (!obj) return undefined

  return obj[ key ]
}
