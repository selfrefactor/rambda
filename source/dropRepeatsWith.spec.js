import { dropRepeatsWith as dropRepeatsWithRamda, eqProps } from 'ramda'

import { compareCombinations } from './_internals/testUtils.js'
import { dropRepeatsWith } from './dropRepeatsWith.js'
import { path } from './path.js'
import { prop } from './prop.js'

const eqI = eqProps('i')

test('happy', () => {
  const list = [ { i : 1 }, { i : 2 }, { i : 2 }, { i : 3 } ]
  const expected = [ { i : 1 }, { i : 2 }, { i : 3 } ]
  const result = dropRepeatsWith(eqI, list)
  expect(result).toEqual(expected)
})

test('readme example', () => {
  const list = [
    {
      a : 1,
      b : 2,
    },
    {
      a : 1,
      b : 3,
    },
    {
      a : 2,
      b : 4,
    },
  ]
  const result = dropRepeatsWith(prop('a'), list)
  expect(result).toEqual([
    {
      a : 1,
      b : 2,
    },
  ])
})

test('keeps elements from the left predicate input', () => {
  const list = [
    {
      i : 1,
      n : 1,
    },
    {
      i : 1,
      n : 2,
    },
    {
      i : 1,
      n : 3,
    },
    {
      i : 4,
      n : 1,
    },
    {
      i : 4,
      n : 2,
    },
  ]
  const expected = [
    {
      i : 1,
      n : 1,
    },
    {
      i : 4,
      n : 1,
    },
  ]
  const result = dropRepeatsWith(eqI)(list)
  expect(result).toEqual(expected)
})

const possiblePredicates = [
  null,
  undefined,
  x => x + 1,
  x => true,
  x => false,
  x => '',
  path([ 'a', 'b' ]),
]
const possibleLists = [
  null,
  undefined,
  [],
  [ 1 ],
  [ { a : { b : 1 } }, { a : { b : 1 } } ],
  [ /foo/g, /foo/g ],
]

describe('brute force', () => {
  compareCombinations({
    firstInput  : possiblePredicates,
    secondInput : possibleLists,
    callback    : errorsCounters => {
      expect(errorsCounters).toMatchInlineSnapshot(`
        {
          "ERRORS_MESSAGE_MISMATCH": 4,
          "ERRORS_TYPE_MISMATCH": 14,
          "RESULTS_MISMATCH": 0,
          "SHOULD_NOT_THROW": 0,
          "SHOULD_THROW": 0,
          "TOTAL_TESTS": 42,
        }
      `)
    },
    fn      : dropRepeatsWith,
    fnRamda : dropRepeatsWithRamda,
  })
})
