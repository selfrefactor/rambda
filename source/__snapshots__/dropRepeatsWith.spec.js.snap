// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`brute force Function | Null 1`] = `
{
  "error": {
    "message": "null is not a list",
    "type": "Error",
  },
  "inputs": [
    [Function],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Null 2`] = `
{
  "error": {
    "message": "null is not a list",
    "type": "Error",
  },
  "inputs": [
    [Function],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Null 3`] = `
{
  "error": {
    "message": "null is not a list",
    "type": "Error",
  },
  "inputs": [
    [Function],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Null 4`] = `
{
  "error": {
    "message": "null is not a list",
    "type": "Error",
  },
  "inputs": [
    [Function],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Null 5`] = `
{
  "error": {
    "message": "null is not a list",
    "type": "Error",
  },
  "inputs": [
    [Function],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Undefined 1`] = `
{
  "error": {
    "message": "undefined is not a list",
    "type": "Error",
  },
  "inputs": [
    [Function],
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of undefined (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Undefined 2`] = `
{
  "error": {
    "message": "undefined is not a list",
    "type": "Error",
  },
  "inputs": [
    [Function],
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of undefined (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Undefined 3`] = `
{
  "error": {
    "message": "undefined is not a list",
    "type": "Error",
  },
  "inputs": [
    [Function],
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of undefined (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Undefined 4`] = `
{
  "error": {
    "message": "undefined is not a list",
    "type": "Error",
  },
  "inputs": [
    [Function],
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of undefined (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Undefined 5`] = `
{
  "error": {
    "message": "undefined is not a list",
    "type": "Error",
  },
  "inputs": [
    [Function],
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of undefined (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Array 1`] = `
{
  "error": {
    "message": "predicate is not a function",
    "type": "TypeError",
  },
  "inputs": [
    null,
    [
      {
        "a": {
          "b": 1,
        },
      },
      {
        "a": {
          "b": 1,
        },
      },
    ],
  ],
  "label": "errors messages are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Array 2`] = `
{
  "error": {
    "message": "predicate is not a function",
    "type": "TypeError",
  },
  "inputs": [
    null,
    [
      /foo/g,
      /foo/g,
    ],
  ],
  "label": "errors messages are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Null 1`] = `
{
  "error": {
    "message": "null is not a list",
    "type": "Error",
  },
  "inputs": [
    null,
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Undefined 1`] = `
{
  "error": {
    "message": "undefined is not a list",
    "type": "Error",
  },
  "inputs": [
    null,
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of undefined (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Undefined | Array 1`] = `
{
  "error": {
    "message": "predicate is not a function",
    "type": "TypeError",
  },
  "inputs": [
    undefined,
    [
      {
        "a": {
          "b": 1,
        },
      },
      {
        "a": {
          "b": 1,
        },
      },
    ],
  ],
  "label": "errors messages are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Undefined | Array 2`] = `
{
  "error": {
    "message": "predicate is not a function",
    "type": "TypeError",
  },
  "inputs": [
    undefined,
    [
      /foo/g,
      /foo/g,
    ],
  ],
  "label": "errors messages are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Undefined | Null 1`] = `
{
  "error": {
    "message": "null is not a list",
    "type": "Error",
  },
  "inputs": [
    undefined,
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Undefined | Undefined 1`] = `
{
  "error": {
    "message": "undefined is not a list",
    "type": "Error",
  },
  "inputs": [
    undefined,
    undefined,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of undefined (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;
